File: quiz_app/count_questions.dart
import 'dart:io';

void main() {
  final files = [
    'lib/screens/local_questions.dart',
    'lib/screens/additional_questions.dart',
    'lib/screens/generated_questions.dart'
  ];

  final Map<String, Map<String, int>> counts = {};

  for (final filePath in files) {
    final file = File(filePath);
    if (!file.existsSync()) {
      print('File not found: $filePath');
      continue;
    }

    final content = file.readAsStringSync();

    // Regex to capture category and difficulty (supporting single or double quotes)
    final categoryRegex =
        RegExp(r'["' ']category["' ']:s*["' ']([^"' ']+)["' ']');
    final difficultyRegex =
        RegExp(r'["' ']difficulty["' ']:s*["' ']([^"' ']+)["' ']');

    // We need to iterate through the file and match pairs.
    // Since it's a list of maps, we can split by "{" to get individual question blocks roughly.
    final blocks = content.split('{');

    for (final block in blocks) {
      final categoryMatch = categoryRegex.firstMatch(block);
      final difficultyMatch = difficultyRegex.firstMatch(block);

      if (categoryMatch != null && difficultyMatch != null) {
        final category = categoryMatch.group(1)!;
        final difficulty = difficultyMatch.group(1)!;

        counts.putIfAbsent(category, () => {'easy': 0, 'medium': 0, 'hard': 0});
        counts[category]![difficulty] =
            (counts[category]![difficulty] ?? 0) + 1;
      }
    }
  }

  print('--- Question Counts ---');
  counts.forEach((category, diffs) {
    print('\nCategory: $category');
    diffs.forEach((diff, count) {
      print('  $diff: $count');
      if (count < 20) {
        print('    [WARNING] Less than 20 questions!');
      }
    });
  });
}

File: quiz_app/count_questions.py
import re
import os

file_path = 'lib/screens/local_questions.dart'

if not os.path.exists(file_path):
    print(f"Error: {file_path} not found.")
    exit(1)

with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
    lines = f.readlines()

count_dict = {}
current_category = None
current_difficulty = None

for line in lines:
    # Match category
    cat_match = re.search(r'"category":\s*"(.*?)"', line)
    if cat_match:
        current_category = cat_match.group(1)
        # Normalize category names if needed (e.g. Science: Computers -> Science)
        # For now, keep exact.

    # Match difficulty
    diff_match = re.search(r'"difficulty":\s*"(.*?)"', line)
    if diff_match:
        current_difficulty = diff_match.group(1)

    # Count at a marker (e.g., correct_answer) to avoid incomplete parsing
    # But difficulty usually comes AFTER category in this file. 
    # Let's count when we hit "question" or "type"?
    # The order seems constant: category, type, difficulty, question.
    # So if we have both and hit "question", we count.
    
    if '"question":' in line and current_category and current_difficulty:
        key = (current_category, current_difficulty)
        count_dict[key] = count_dict.get(key, 0) + 1
        # Reset is tricky if order varies, but let's assume valid blocks reset logic naturally
        # Actually, in a map, fields are separate.
        # We assume one question per block.
        # We can just reset after counting.
        current_category = None
        current_difficulty = None

print(f"{'Category':<30} | {'Difficulty':<10} | {'Count'}")
print("-" * 55)

categories = sorted(list(set(k[0] for k in count_dict.keys())))
difficulties = ['easy', 'medium', 'hard']

low_buckets = []

for cat in categories:
    for diff in difficulties:
        count = count_dict.get((cat, diff), 0)
        print(f"{cat:<30} | {diff:<10} | {count}")
        if count < 20:
            low_buckets.append((cat, diff, 20 - count))

print("\n--- DEFICITS (< 20) ---")
for cat, diff, missing in low_buckets:
    print(f"{cat} - {diff}: Need {missing} more")
